### PR TITLE
Fix hang in graph with absolute point as offset

### LINKED ISSUE #1394: Use of absolute cycle points in graph offsets.
```
[scheduling]
    initial cycle point = 1
    cycling mode = integer
    [[dependencies]]
        [[[R1/+P1]]]
            graph = start
        [[[R//P1]]]
            graph = start[^+P1] => foo   # good
            # graph = start[2] => foo  # BAD
```

The two variants of the inter-cycle trigger above should be equivalent.  The absolute version graphs correctly but on validation it gets stuck in an endless loop.  A quick look at this suggests it's computing the cleanup cut-off point that's causing the problem.

### PR DESCRIPTION
**Pull request by matthewrmshin**
Improve cutoff point for absolute offset.
* Use stop point of sequence as cutoff point.

Improve error message for bad recurrence syntax
Allow long date-time format in graph offset.
Document performance issue with immortal fixed cycle point tasks.

Fix #1394 (and #1951).

### CODE REVIEW

File: lib/cylc/cycling/__init__.py [hunk 1]
```diff
@@ -127,6 +127,8 @@ def __str__(self):
         # Stringify.
         return self.value
 
+    __repr__ = __str__
```
Comment (matthewrmshin) [inline lib/cylc/cycling/__init__.py]: Friendlier display in `print` statements.

**Comment by oliver-sanders** [comment]
The example suite in #1394 will stall if run as the start task is not retained in the pool.

**Comment by oliver-sanders** [comment]
Needs a test to ISO8601 absolute points. The following suite fails validation with `'NoneType' object has no attribute 'set_time_zone'` due to the `foo[20000102T00Z]` bit.

```
[cylc]
    UTC mode = True
[scheduling]
    initial cycle point = 20000101T00Z
    [[dependencies]]
        [[[R1]]]
            graph = start
        [[[T00]]]
            graph = start[^] => foo
        [[[20000102T12Z]]]
            graph = foo[20000102T00Z] => bar
```

**Comment by matthewrmshin** [comment]
> The example suite in #1394 will stall if run as the start task is not retained in the pool.

(Looks like we have discovered even more badness in the code.)

File: tests/validate/59-absolute-as-offset.t [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    [[[T12]]]
+        graph = fixed_cycle[20100101T0600] => foo
+
+[runtime]
+    [[foo, fixed_cycle, init_cycle]]
+        script = true
+__SUITE_RC__
+
+run_ok "${TEST_NAME_BASE}-datetime" cylc validate 'suite.rc'
+
```
Comment (arjclark) [inline tests/validate/59-absolute-as-offset.t]: Test should actually run the suite to ensure it is working

**Comment by matthewrmshin** [comment]
Now running the previously bad suites, instead of just validating them.

**Comment by matthewrmshin** [comment]
The `'NoneType' object has no attribute 'set_time_zone'` problem was a syntax error. The `20000102T12Z` section needs to be `R1/20000102T12Z`. I have added a better error message for this.

**Comment by matthewrmshin** [comment]
Should all be good now.

**Comment by matthewrmshin** [comment]
(See also #1773.)

**Comment by hjoliver** [comment]
(tests as working, btw).

File: doc/cug.tex [hunk 3]
```diff
@@ -3720,6 +3720,34 @@ \subsubsection{Trigger Types}
...
+\begin{lstlisting}
+[scheduling]
+    initial cycle point = 2010
+    [[dependencies]]
+        # This is BAD!
+        [[[P1D]]]
+            graph = "baz[20200101] => qux"
+\end{lstlisting}
+\lstset{language=transcript}
+
```
Comment (matthewrmshin) [inline doc/cug.tex]: Yes, this is the same as `baz[^] => qux` in a cycling section (see #1773). I'll explain the performance issue in the docs for now.

**Comment by oliver-sanders** [comment]
Note this will currently fail for some of the more exotic [ISO8601 datetime formats](https://github.com/metomi/isodatetime/blob/master/README.md):
- 2000-01-01T0000
- 20000101T00:00
- 2000-W01-01T0000
- 2000-001T0000
- 20000101T00,0
- 20000101T00.0
- 20000101T0000-00:00

**Comment by hjoliver** [comment]
It should presumably accept the configured `cycle point format` of the suite, at the least.

File: lib/cylc/config.py [hunk 4]
```diff
@@ -1822,9 +1822,14 @@ def load_graph(self):
...
-                               self.cfg['scheduling']['initial cycle point'],
-                               self.cfg['scheduling']['final cycle point'])
+            try:
+                seq = get_sequence(
+                    section,
+                    self.cfg['scheduling']['initial cycle point'],
+                    self.cfg['scheduling']['final cycle point'])
+            except (AttributeError, TypeError):
+                raise SuiteConfigError(
+                    "ERROR: Invalid recurrence representation: %s" % section)
```
Comment (matthewrmshin) [inline lib/cylc/config.py]: Looking at the statement in the `try:` block above, the argument to the `get_sequence` function are the section title (which should be a cycle point recurrence), the initial cycle point and the final cycle point.

The initial and final cycle points should already have their syntaxes checked earlier, so the only thing that can cause exception here should be invalid or unsupported recurrence syntax, or a bug in our system.

**Comment by matthewrmshin** [comment]
Maybe even more badness in the date-time parsing logic?

**Comment by hjoliver** [comment]
Can you consult with Ben over morning tea? 😄

**Comment by matthewrmshin** [comment]
Actually, I cannot see how any of the above can work.
* 2000-01-01T0000
  * mixed format - long format for date and short format for time
* 20000101T00:00
  * mixed format - short format for date and long format for time
* 2000-W01-01T0000
  * syntax error: day of week should be single digit
* 2000-001T0000
  * mixed format
* 20000101T00,0 and 20000101T00.0
  * any usage for these?
* 20000101T0000-00:00
  * mixed format - time zone in long format, the rest in short format

Note: section header appears to be OK with long format or short format, but not both. For offset, the regular expression appears to only support the short format - as the colon `:` character is not included. I'll see if we have any badness for supporting the colon character in offset, but maybe this is better for a different issue?

**Comment by hjoliver** [comment]
I guess those formats were a bit too exotic 😬

**Comment by matthewrmshin** [comment]
I have now added the colon `:` character to the regular expression for offset. It does not appear to cause any badness yet.

**Comment by matthewrmshin** [comment]
Branch squashed and rebased.

**Comment by arjclark** [comment]
@oliver-sanders - given all your recent testing etc. of this can you do the review 2 of this please.

**Pull request merged**